subsys/bootloader/CMakeLists.txt

```diff
@@ -171,19 +171,92 @@ include (${CMAKE_CURRENT_LIST_DIR}/cmake/debug_keys.cmake)
 
 set(sign_depends kernel_elf ${key_file_depends})
 
+set(hash_file firmware.sha256)
+set(signature_file firmware.signature)
+set(public_key_file public.pem)
+
+set(hashcmd
+  ${PYTHON_EXECUTABLE}
+  ${NRF_BOOTLOADER_SCRIPTS}/hash.py --in ${PROJECT_BINARY_DIR}/${KERNEL_BIN_NAME} > ${hash_file}
+  )
+
+if (${CONFIG_SB_SIGNING_PYTHON})
+  set(signcmd
+    ${PYTHON_EXECUTABLE}
+    ${NRF_BOOTLOADER_SCRIPTS}/do_sign.py --private-key ${signature_private_key_file} --in ${hash_file} > ${signature_file}
+    )
+  set(pubgencmd
+    ${PYTHON_EXECUTABLE}
+    ${NRF_BOOTLOADER_SCRIPTS}/keygen.py --public --in ${signature_private_key_file} --out ${public_key_file}
+    )
+elseif (${CONFIG_SB_SIGNING_OPENSSL})
+  set(signcmd
+    openssl dgst -sha256 -sign ${signature_private_key_file} ${hash_file} |
+    ${PYTHON_EXECUTABLE}
+    ${NRF_BOOTLOADER_SCRIPTS}/asn1parse.py --alg ecdsa --contents signature > ${signature_file}
+    )
+  set(pubgencmd
+    openssl ec -pubout -in ${signature_private_key_file} -out ${public_key_file}
+    )
+elseif (${CONFIG_SB_SIGNING_CUSTOM})
+  set(signcmd
+    ${CONFIG_SB_SIGNING_COMMAND} ${hash_file} > ${signature_file}
+    )
+  set(public_key_file ${CONFIG_SB_SIGNING_PUBLIC_KEY})
+  if (signcmd STREQUAL "" OR NOT EXISTS ${public_key_file})
+    message(WARNING "You must specify a signing command and valid public key file.")
+  endif()
+else ()
+  message(WARNING "Unable to parse signing config.")
+endif()
+
+add_custom_command(
+  OUTPUT
+  ${signature_file}
+  COMMAND
+  ${hashcmd}
+  COMMAND
+  ${signcmd}
+  DEPENDS
+  ${sign_depends}
+  WORKING_DIRECTORY ${PROJECT_BINARY_DIR}
+  COMMENT
+  "Creating signature of application."
+  USES_TERMINAL
+  )
+
+if (${CONFIG_SB_PRIVATE_KEY_HANDLING})
+  add_custom_command(
+    OUTPUT
+    ${public_key_file}
```

Comment:
if private key handling, create a public key file? What is the logic here, and the rationale for the kconfig name?

Reply:
I'd like to come up with a better name. `SB_PRIVATE_KEY_HANDLING` is for when the private key is available to to be handled by the build system. Otherwise, the public key must be provided separately, in which case this step is not needed.

Reply:
Changed the name to `SB_PRIVATE_KEY_PROVIDED`